src/tiled/stampbrush.cpp

```diff
@@ -386,6 +401,52 @@ void StampBrush::drawPreviewLayer(const QVector<QPoint> &list)
             mapDocument()->unifyTilesets(variation.map, mMissingTilesets);
 
             TileLayer *stamp = variation.tileLayer();
+            bool copyFlag = false;
+
+            //if staggered map, makes sure stamp stays the same
+            if ((mapDocument()->map()->orientation() == Map::Hexagonal
+                    || mapDocument()->map()->orientation() == Map::Staggered)
+                    && ((mapDocument()->map()->staggerAxis() == Map::StaggerY)?
+                         stamp->height() > 1 : stamp->width() > 1)) {
+                stamp = variation.tileLayer()->copy(variation.tileLayer()->bounds());
+                copyFlag = true;
```

Comment:
At this point you still don't know whether it is actually necessary to create a copy. Also, you only ever need a single adjusted copy, since there is only a single possible adjustment depending on the context. So I think you can really just store it in a `QScopedPointer<TileLayer*> copy` (or maybe call it `shiftedCopy`) and create it on-demand.